File: hw1_wma/Nature.m
classdef Nature
    %UNTITLED Summary of this class goes here
    %   Detailed explanation goes here
    
    properties
        observation = []
        label = 0
        e_prediction = []
        final_prediction = 0
        weights = []
    end
    
    methods
        %% Get functions
        % get the observation
        function obs = get_observation(obj)
            obs = obj.observation;
        end
        % Get the label
        function lab = get_label(obj)
            lab = obj.label;
        end
        function [weights] = get_weights(obj)
            weights = obj.weights;
        end
        %% Set functions
        % set the expert predictions we get from learner
        function set_expert_prediction(obj, e_pred)
            obj.e_prediction = e_pred;
        end
        % set the final prediction
        function set_final_prediction(obj, pred)
            obj.final_prediction = pred;
        end
        % set the weights of the learner
        function set_weights(obj, wts)
            obj.weights = wts;
        end
    end
    
end

File: hw1_wma/Nature.cpp
#include "Nature.h"

Nature::Nature() {
}

Nature::Nature(int rounds) {
    srand(time(NULL));
    for(int i=0; i<rounds; i++) {
        std::vector<int> temp;
        temp.push_back(i);          // the game num
        temp.push_back(rand()%2);   // weather = sunny?
        temp.push_back(rand()%2);   // field = grass?
        temp.push_back(rand()%2);   // location = home?
        vec_observations.push_back(temp);
    }
}

Nature::~Nature() {
}

/***********************************************************************
 * Get the label.
 */
int Nature::get_label(int round) {
    // Calculate the label
    calculate_label(round);

    return label;
}

/***********************************************************************
 * Get the label.
 * In the adversarial case, we'll need to look at the weights and expert
 * predictions of the Learner.
 */
int Nature::get_label(int round, std::vector<int> e_pred, std::vector<double> wts) {
    vec_e_predictions = e_pred;
    vec_weights = wts;

    // Calculate the label
    calculate_label(round);

    return label;
}

/***********************************************************************
 * Return the observations of the given round.
 */
std::vector<int> Nature::get_observations(int round) {
    return vec_observations[round];
}

/***********************************************************************
 *
 */
void Nature::set_expert_predictions(std::vector<int> e_pred) {
    vec_e_predictions = e_pred;
}

/***********************************************************************
 *
 */
void Nature::set_final_prediction(int pred) {
    prediction = pred;
}

/***********************************************************************
 *
 */
void Nature::set_weights(std::vector<double> wts) {
    vec_weights = wts;
}
